articles/clojure_elixir_fibonacci_mstalker/article.md

```diff
@@ -0,0 +1,339 @@
+
+That solution works, but it's arguably not very idiomatic Elixir. Let's try again.
+
+```elixir
+defmodule Fibonacci do
+  def compute(0), do: 0
+  def compute(1), do: 1
+  def compute(n), do: compute(n - 1) + compute(n - 2)
+end
+```
+
+This time, I wrote three different function clauses. This takes advantage of Elixir's pattern-matching. When you call the `compute/1` function, Elixir will look at each function clause from top to bottom, testing each line to see whether a given function clause will match the argument. If `n` is `0`, Elixir sees that the first function clause is meant to work with an argument of `0`, and it executes the function clause. This function returns `0`.
+
+If `n` is `1`, Elixir tries to match this argument with the first function clause, and fails. It then tries to match `n` with the `1` in the signature of the second function clause. There's a match, so `compute/1` returns `1`.
+
+If `n` is greater than `1`, Elixir executes the last function clause, since the first two function clauses only match `0` or `1`, and the last function clause matches any argument value.
+
+We could shorten our function even more by adding a [guard clause](https://hexdocs.pm/elixir/guards.html#content):
+
+```elixir
+defmodule Fibonacci do
+  def compute(n) when n in [0, 1], do: n
+  def compute(n), do: compute(n - 1) + compute(n - 2)
+end
+```
+
+A guard clause adds a condition to the function clauses. If the pattern of the argument matches the function signature, and the arguments satisfy the guard clause, the function executes; otherwise, it doesn't. In this case, I used a membership guard called [`in`](https://hexdocs.pm/elixir/Kernel.html#in/2). If `n` is equal to any element in the list `[0, 1]`, Elixir executes the first function clause and returns `n`. Otherwise, Elixir executes the second function clause.
+
+
+## Optimizing Our Functions
+
+The functions above are fine for small values of `n`. If `n` is sufficiently large, though, these functions will consume a large quantity of memory and will slow down significantly. The Clojure version even raises a `StackOverflow Error`. Try calling these functions when `n` is `200000` and you'll see what I mean!
+
+So far, our functions figured out a value by starting at a higher sequence index, and then calculating the values at lower indexes. We can speed up our function execution and lower memory consumption by starting at lower indexes and working our way up. Writing functions this way tends to be more complicated because they require one or more state variables that we didn't need in our simpler implementations above.
+
+
+### Clojure
+
+Let's take a look at a Clojure implementation.
+
+```clojure
+(ns fibonacci)
+
+(defn- helper
+  [n]
+  (loop [current-index 2
+         current-value 1
+         previous-value 1]
+    (if (= n current-index)
+      current-value
+      (recur (inc current-index)
```

Comment:
Is it necessary (or somehow better) to use `inc` to increment a number in clojure?
Why not just `(+ current-index 1)` ?

Reply:
`(inc current-index)` and `(+ current-index 1)` are equivalent. I figured that since Clojure offers an `inc` function, I'd use it, rather than write out the addition.